x-pack/plugins/lens/public/shared_components/coloring/color_stops.tsx

```diff
@@ -0,0 +1,285 @@
+}: CustomStopsProps) => {
+  const onChangeWithValidation = useCallback(
+    (newColorStops: Array<{ color: string; stop: string }>) => {
+      const areStopsValid = newColorStops.every(({ color, stop }, i) => {
+        const numberStop = Number(stop);
+        const prevNumberStop = Number(newColorStops[i - 1]?.stop ?? -Infinity);
+        return isValidColor(color) && !Number.isNaN(numberStop) && numberStop >= prevNumberStop;
+      });
+      if (areStopsValid) {
+        onChange(newColorStops.map(({ color, stop }) => ({ color, stop: Number(stop) })));
+      }
+    },
+    [onChange]
+  );
+  const memoizedValues = useMemo(
+    () => colorStops.map(({ color, stop }) => ({ color, stop: String(stop), id: idGenerator() })),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [colorStops, reverse]
+  );
+  const { inputValue: localColorStops, handleInputChange: setLocalColorStops } = useDebouncedValue({
+    onChange: onChangeWithValidation,
+    value: memoizedValues,
+  });
+  const [sortedReason, setSortReason] = useState<string>('');
+  const shouldEnableDelete = localColorStops.length > 2;
+
+  const [popoverInFocus, setPopoverInFocus] = useState<boolean>(false);
+
+  // refresh on unmount:
+  // the onChange logic here is a bit different than the one above as it has to actively sort if required
+  useUnmount(() => {
+    const areStopsValid = localColorStops.every(({ color, stop }, i) => {
+      const numberStop = Number(stop);
+      return isValidColor(color) && !Number.isNaN(numberStop);
+    });
+    const shouldSort = localColorStops.some(({ stop }, i) => {
+      const numberStop = Number(stop);
+      const prevNumberStop = Number(localColorStops[i - 1]?.stop ?? -Infinity);
+      return numberStop < prevNumberStop;
+    });
+    if (areStopsValid && shouldSort) {
+      onChange(
+        localColorStops
+          .map(({ color, stop }) => ({ color, stop: Number(stop) }))
+          .sort(({ stop: stopA }, { stop: stopB }) => Number(stopA) - Number(stopB))
+      );
+    }
+  });
+
+  return (
+    <>
+      {sortedReason ? (
+        <EuiScreenReaderOnly>
+          <p aria-live="assertive">
+            {i18n.translate('xpack.lens.dynamicColoring.customPalette.sortReason', {
+              defaultMessage: 'Color stops have been sorted due to new stop value {value}',
+              values: {
+                value: sortedReason,
+              },
+            })}
+          </p>
+        </EuiScreenReaderOnly>
+      ) : null}
+
+      <EuiFlexGroup
+        data-test-subj={`${dataTestPrefix}_dynamicColoring_custom_stops`}
+        direction="column"
+        gutterSize="s"
+      >
+        {localColorStops.map(({ color, stop, id }, index) => {
+          const prevStopValue = Number(localColorStops[index - 1]?.stop ?? -Infinity);
+          const nextStopValue = Number(localColorStops[index + 1]?.stop ?? Infinity);
+
+          return (
+            <EuiFlexItem
+              key={id}
```

Comment:
There's still something not workiing completely right - I think it's because the id is not persisted between rerenders.
What happens is that when I type a number the focus is being lost IF the number change won't cause reordering. 
Try for example with ranges [20,40,60,80] to change 20 to 2 and then try to add zero again.